veps/sig-compute/tdx-sev-snp-proposal.md

```diff
@@ -0,0 +1,236 @@
+
+```xml
+   spec:
+     domain:
+       launchSecurity:
+         tdx: {}
+```
+Let's hold off on adding more options to the VMI spec for now. We need further
+discussion on the use cases and how they would apply to VM creation.
+
+### AMD Design
+
+AMD SEV-SNP maps to a new type in the LibVirt XML of type “sev-snp” with its
+own attributes and elements that are similar to the type “sev”.
+
+```xml
+   <domain>
+     ...
+     <launchSecurity type='sev-snp'>
+       <cbitpos>47</cbitpos>
+       <policy>0x00030000</policy>
+       <reducedPhysBits>1</reducedPhysBits>
+	<guestVisibileWorkarounds>...</guestVisibleWorkaround>
+	<idBlock>...</idBlock>
+	<idAuth>...</idAuth>
+	<hostData>...</hostData>
+     </launchSecurity>
+     ...
+   </domain>
+```
+
+A new structure should be created for adding SEV-SNP on the `LaunchSecurity`
+field.
+```go
+type SEVSNP struct {
+  // 64-bit SEV-SNP Guest Policy
+  policy: "0x3000" # Default Policy
+}
+```
+Similar to TDX, all elements for type=’sev-snp’, this will be a basic enablement
+until further use cases arise to provide additional settings to configure a 
+Confidential VM. The default policy option will have a default value that QEMU specifies.
+The node labeller will detect SEV-SNP capabilities from the LibVirt domain capabilities then
+apply the label to the node, while the node selector renderer will be extended to include
+SEV-SNP scheduling.
+
+### Security Considerations
+- The infrastructure provider and the VMI author are untrusted. For example,
+  the infrastructure provider is always capable of injecting random cloud-init
+  stuff behind our back. So if an instance of a secret application runs inside
+  a CVM, via cloud-init, an infrastructure provider or a hacker could simply
+  execute random commands inside this VM.
+
+## API Examples
+The bellow yaml snippets provide examples of how to request TDX feature in the
+VMI spec.
+
+### API Examples for Intel TDX
+
+- TDX VM with default options
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  ...
+spec:
+  domain:
+    launchSecurity:
+      tdx: {}
+  ...
+```
+
+### API Examples for AMD
+
+- Basic SEV-SNP with default options
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: sev-snp-vm
+spec:
+  domain:
+    launchSecurity:
+      snp: {}
+  ...
+```
+
+## Alternatives
+
+## Scalability
+
+## Update/Rollback Compatibility
+- All new fields are fields should be optional
+- This should not impact existing VMs
+- AMD SEV Compatibility:
+  - The AMD SEV-SNP feature sits behind the existing SEV feature gate without
+    breaking changes
+  - Node labels are added, no labels are removed.
+
+## Functional Testing Approach
+- Unit testing to detect TDX & SEV-SNP from the libvirt capabilities.
+- Since TDX & SEV-SNP does not have support for nested virtualization this will
+  require bare metal hardware to conduct e2e testing.
```

Comment:
Do we have plans to acquire such hardware? Will it be available before the end of the implementation phase? Also, integrating the hardware to CI and creating test lanes should probably be an implementation phase.
```suggestion
- Since TDX & SEV-SNP do not have support for nested virtualization this will
  require bare metal hardware to conduct e2e testing.
```

Reply:
> Do we have plans to acquire such hardware? 

I'm working on this for TDX.